floodlight/io/kinexon.py

```diff
@@ -0,0 +1,359 @@
+        Can be one of:
+            - 'sensor_id'
+            - 'mapped_id'
+            - 'name'
+        When recording and exporting Kinexon data, the pID can be stored
+        in different columns. Player-identifying columns are "sensor_id", "mapped_id",
+        and "full_name". If specified to one of the above, keys in links will be the
+        pIDs in that column. If not specified, it will use one of the columns, favoring
+        "name" over "mapped_id" over "sensor_id".
+    Returns
+    -------
+    links: Dict[str, Dict[str, int]]
+        Link-dictionary of the form `links[group][identifier-ID] = xID`.
+    """
+
+    if identifier is None:
+        for player_id in ["name", "mapped_id", "sensor_id"]:
+            if player_id in list(pID_dict.values())[0]:
+                identifier = player_id
+                break
+
+    links = {}
+    for gID in pID_dict:
+        links.update(
+            {gID: {ID: xID for xID, ID in enumerate(pID_dict[gID][identifier])}}
+        )
+
+    return links
+
+
+def read_kinexon_file(filepath_data: Union[str, Path]) -> List[XY]:
+    """Parse Kinexon files and extract position data.
+
+    Kinexon's local positioning system delivers one .csv file containing the position
+    data. This function provides a high-level access to Kinexon data by parsing "the
+    full file" given the path to the file.
+
+    Parameters
+    ----------
+    filepath_data: str of pathlib.Path
+        Full path to Kinexon.csv-file.
+
+    Returns
+    -------
+    positions: List[XY]
+        XY-Object for the whole game. The order of groups is ascending according to
+        their group_id. The order inside the groups is ascending according to their
+        appearance in the data.
+    """
+
+    # get metadata
+    pID_dict, number_of_frames, framerate, t_null = get_meta_data(filepath_data)
+
+    # get links
+    links = create_links_from_meta_data(pID_dict)
+    # get column-links
+    column_links = _get_column_links(filepath_data)
+    column_links_set = set(column_links)
+    group_identifier_set = {"group_id", "group_name"}
+    recorded_group_identifier = list(column_links_set & group_identifier_set)
+
+    # create np.array for data
+    number_of_sensors = {}
+    xydata = {}
+    for group in links:
+        number_of_sensors.update({group: len(links[group])})
+        xydata.update(
+            {
+                group: np.full(
+                    [number_of_frames + 1, number_of_sensors[group] * 2], np.nan
+                )
+            }
+        )
+
+    has_groups = len(recorded_group_identifier) > 0
+    if has_groups:
+        if "group_name" in recorded_group_identifier:
+            group_identifier = "group_name"
+        elif "group_id" in recorded_group_identifier:
+            group_identifier = "group_id"
+
+        def get_group_id(single_line):
+            return single_line[column_links[group_identifier]]
+
+    # no groups
+    else:
+
+        def get_group_id(single_line):
```

Comment:
Although I like your solution to dynmaic group id creation, nested functions are not the most stable way. I've also noticed you defined `get_group_id` four times throughout your code. Maybe there is a way to refactor all four into one top level function `_get_group_id(single_line)` that returns `single_line[column_links[group_identifier]]` if the value exists, or "0" otherwise.

If I'm missing something and thats not straightforward to solve, I'm fine going with your solution!